### PR TITLE
First draft of a device location widget

### PR DESCRIPTION
**Pull request by esundberg**
DO NOT DELETE THIS TEXT

#### Please note

> Please read this information carefully. You can run `./scripts/pre-commit.php` to check your code before submitting.

- [ ] Have you followed our [code guidelines?](http://docs.librenms.org/Developing/Code-Guidelines/)

#### Testers

If you would like to test this pull request then please run: `./scripts/github-apply <pr_id>`, i.e `./scripts/github-apply 5926`
After you are done testing, you can remove the changes with `./scripts/github-remove`.  If there are schema changes, you can ask on discord how to revert.

### CODE REVIEW

File: app/Http/Controllers/Widgets/DeviceLocationController.php [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+    #{
+    #
+    #}
+
+    public function getView(Request $request)
+    {
+
+        $data = $this->getSettings();
+
+        $data['locations'] = DB::select("select
```
Comment (murrant) [inline app/Http/Controllers/Widgets/DeviceLocationController.php]: This should work

```php
$data['locations'] = Location::orderBy('location')->get()->map(function ($location) {
    /** @var Location $location */
    return [
        'location' => $location->location,
        'total' => $location->devices()->count(),
        'up' => $location->devices()->isUp()->count(),
        'down' => $location->devices()->isDown()->count(),
        'ignored' => $location->devices()->isIgnored()->count(),
        'disabled' => $location->devices()->isDisabled()->count(),
    ];
});
```

I'm not sure the totals belongs in this widget... there are other ones that do that.

Basically you can use the same queries as above for the totals if you want them.
I'm also not sure why you have included the id.

Comment (esundberg) [reply app/Http/Controllers/Widgets/DeviceLocationController.php]: @murrant That worked thanks.

How do I add a row to the widget table?      
Need to add the following to the widget table. But i think it's seeded from somewhere else.
widget_id=17 | Device Location           | device-location

Comment (esundberg) [reply app/Http/Controllers/Widgets/DeviceLocationController.php]: @murrant bump.... looking to know how to add a row to the widgets table.

Comment (murrant) [reply app/Http/Controllers/Widgets/DeviceLocationController.php]: @esundberg you will need to create a [database migration](https://laravel.com/docs/migrations) and update the seeder database/seeds/DefaultWidgetSeeder.php

Comment (esundberg) [reply app/Http/Controllers/Widgets/DeviceLocationController.php]: Ok thanks I will get to this shortly

**Comment by murrant** [comment]
Web files have been moved to includes/html, you will need to update your PR, sorry for the trouble. Let us know if you need any help.

File: app/Http/Controllers/Widgets/DeviceLocationController.php (hunk 1 above)
Comment (esundberg) [reply app/Http/Controllers/Widgets/DeviceLocationController.php]: I ran into an issue while running this in our test environment. Need to make some correction. It's displaying locations that have no devices associated to it.

Comment (esundberg) [reply app/Http/Controllers/Widgets/DeviceLocationController.php]: @murrant It looks like the location's are not being removed and always being added. How would if make the query so it only returns the location if there a 1 or more devices at that location.

Location::orderBy('location') WHERE count > 0

Comment (murrant) [reply app/Http/Controllers/Widgets/DeviceLocationController.php]: Location->whereHas('device')->orderBy('location')->get();

Comment (esundberg) [reply app/Http/Controllers/Widgets/DeviceLocationController.php]: Ok thanks. i will get this push through when i have some time.

**Comment by laf** [comment]
@esundberg Have you managed to get the additional seed update done so we can get this merged?

**Comment by Jellyfrog** [comment]
Closing due to inactivity. Feel free to reopen!

**Pull request closed without merging**